ethcore/src/client/client.rs

```diff
@@ -1853,21 +1853,96 @@ impl BlockChainClient for Client {
 	}
 
 	fn logs(&self, filter: Filter) -> Vec<LocalizedLogEntry> {
-		let (from, to) = match (self.block_number_ref(&filter.from_block), self.block_number_ref(&filter.to_block)) {
-			(Some(from), Some(to)) => (from, to),
-			_ => return Vec::new(),
-		};
+		macro_rules! return_empty_if_none {
+			( $v: expr ) => (
+				match $v {
+					Some(value) => value,
+					None => return Vec::new(),
+				}
+			)
+		}
 
 		let chain = self.chain.read();
-		let blocks = filter.bloom_possibilities().iter()
-			.map(move |bloom| {
-				chain.blocks_with_bloom(bloom, from, to)
-			})
-			.flat_map(|m| m)
-			// remove duplicate elements
-			.collect::<HashSet<u64>>()
-			.into_iter()
-			.collect::<Vec<u64>>();
+
+		// First, check whether `filter.from_block` and `filter.to_block` is on the canon chain. If so, we can use the
+		// optimized version.
+		let is_canon = |id| {
+			match id {
+				// If it is referred by number, then it is always on the canon chain.
+				&BlockId::Earliest | &BlockId::Latest | &BlockId::Number(_) => Some(true),
+				// If it is referred by hash, we see whether a hash -> number -> hash conversion gives us the same
+				// result.
+				&BlockId::Hash(hash) =>
```

Comment:
1. I would move `is_canon(H256) -> bool` to `blockhain.rs` (it's also used in couple of places inside that file (epochs)).
2. This would become just `chain.is_canon(self.block_hash(id)?)`